File: 5. Longest Palindromic Substring/cpp/solution.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int left = 0;
        int right = 0;
        for(int i = 0; i < s.size() - 1; i++){
            int r = i;
            int l = i;
            while(r + 1 < s.size() && s[l] == s[r + 1]){
                r++;
            }
            while(l >= 0 && r < s.size() && s[l] == s[r]){
                if(r - l > right - left){
                    right = r;
                    left = l;
                }
                r++;
                l--;
            }
        }
        string solution = "";
        for(int i = left; i <= right; i++){
            solution+=s[i];
        }
        return solution;
    }
};

File: 5. Longest Palindromic Substring/python/solution.py
class Solution(object):

    def longestPalindrome(self, s):
        left, right = 0, 0
        for i in range(len(s)):
            l1, r1 = self.expandingString(s, i, i + 1)
            l2, r2 = self.expandingString(s, i, i)

            if r1 - l1 >= right - left:
                left = l1
                right = r1
            if r2 - l2 >= right - left:
                left = l2
                right = r2
        return s[left:right + 1]
    def expandingString(self, s, l, r):
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l-=1
            r+=1
        return l+1, r-1